public/app/features/alerting/unified/utils/rulerClient.ts

```diff
@@ -157,65 +141,105 @@ export function getRulerClient(rulerConfig: RulerDataSourceConfig): RulerClient
+      throw new Error(`No group found with name "${group.name}"`);
+    }
 
     const payload: PostableRulerRuleGroupDTO = {
-      name: groupName,
-      interval: evaluateEvery,
-      rules: [formRule],
+      name: group.name,
+      interval: group.interval,
+      rules: (existingGroup.rules ?? []).concat(newRule as RulerGrafanaRuleDTO),
     };
-    await setRulerRuleGroup(rulerConfig, folder.title, payload);
 
-    // now refetch this group to get the uid, hah
-    const result = await fetchRulerRulesGroup(rulerConfig, folder.title, groupName);
-    const newUid = (result?.rules[0] as RulerGrafanaRuleDTO)?.grafana_alert?.uid;
-    if (newUid) {
-      // if folder has changed, delete the old one
-      if (existing) {
-        const freshExisting = await findEditableRule(ruleId.fromRuleWithLocation(existing));
-        if (freshExisting && freshExisting.namespace !== folder.title) {
-          await deleteRule(freshExisting);
-        }
-      }
+    await setRulerRuleGroup(rulerConfig, namespace, payload);
 
-      return { uid: newUid, ruleSourceName: 'grafana' };
-    } else {
-      throw new Error('Failed to fetch created rule.');
+    // TODO figure out how to get the newly created UID :(
+    return { uid: '', ruleSourceName: GRAFANA_RULES_SOURCE_NAME };
+  };
+
+  // we can't move the rule in a single atomic operation so we have to
+  // 1. add the rule to the new group
+  // 2. remove the rule from the old one
+  const moveGrafanaRule = async (
+    namespace: string,
+    group: { name: string; interval: string },
+    existingRule: RuleWithLocation,
+    newRule: PostableRuleGrafanaRuleDTO
+  ): Promise<RuleIdentifier> => {
+    // add the new rule to the requested namespace and group
+    const identifier = await addRuleToNamespaceAndGroup(namespace, group, newRule);
+
+    // remove the rule from the previous namespace and group
+    await deleteRule({
+      ruleSourceName: existingRule.ruleSourceName,
+      namespace: existingRule.namespace,
+      group: existingRule.group,
+      rule: newRule as RulerGrafanaRuleDTO,
+    });
+
+    return identifier;
+  };
+
+  const updateGrafanaRule = async (
+    existingRule: RuleWithLocation,
+    newRule: PostableRuleGrafanaRuleDTO,
+    interval: string
+  ): Promise<RuleIdentifier> => {
+    // type guard to make sure we're working with a Grafana managed rule
+    if (!isGrafanaRulerRule(existingRule.rule)) {
+      throw new Error('The rule is not a Grafana managed rule');
     }
+
+    // make sure our updated alert has the same UID as before
+    const uid = existingRule.rule.grafana_alert.uid;
+    newRule.grafana_alert.uid = uid;
+
+    // create the new array of rules we want to send to the group
+    const newRules = existingRule.group.rules
+      .filter((rule): rule is RulerGrafanaRuleDTO => isGrafanaRulerRule(rule))
+      .filter((rule) => rule.grafana_alert.uid !== uid)
+      .concat(newRule as RulerGrafanaRuleDTO); // TODO why type cast this?
```

Comment:
TODO

Reply:
It was being typecast before, I didn't dig too deep in to the reason why – might dig a little deeper if I have the time.